core/adapters/eth_tx.go

```diff
@@ -42,6 +46,113 @@ func (e *EthTx) TaskType() models.TaskType {
 // is not currently pending. Then it confirms the transaction was confirmed on
 // the blockchain.
 func (e *EthTx) Perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	if store.Config.EnableBulletproofTxManager() {
+		return e.perform(input, store)
+	} else {
+		return e.legacyPerform(input, store)
+	}
+}
+
+// TODO(sam): https://www.pivotaltracker.com/story/show/173280188
+func (e *EthTx) perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	trtx, err := store.FindEthTaskRunTxByTaskRunID(input.TaskRunID().UUID())
+	if err != nil {
+		err = errors.Wrap(err, "FindEthTaskRunTxByTaskRunID failed")
+		logger.Error(err)
+		return models.NewRunOutputError(err)
+	}
+	if trtx != nil {
+		return e.checkForConfirmation(*trtx, input, store)
+	}
+	return e.insertEthTx(input, store)
+}
+
+func (e *EthTx) checkForConfirmation(trtx models.EthTaskRunTx, input models.RunInput, store *store.Store) models.RunOutput {
+	switch trtx.EthTx.State {
+	case models.EthTxConfirmed:
+		return checkEthTxForReceipt(trtx.EthTx.ID, input, store)
+	case models.EthTxFatalError:
+		return models.NewRunOutputError(trtx.EthTx.GetError())
+	default:
+		return models.NewRunOutputPendingOutgoingConfirmationsWithData(input.Data())
```

Comment:
This is a legacy hangover from the way the runmanager works... It's a bit wild but it calls into `Perform` on every new head. I opted to keep it that way for now and will fix later.

It could totally get here in `unstarted` or `in_progress` - in either case it's still ok to be in `PendingOutgoingConfirmations`. We just have to wait for it to get to either `fatal_error` or `unconfirmed/confirmed`

Reply:
The run manager is consuming this output, to know whether its safe to go onto the next task yet.

Reply:
Got it!